src/hazelcore/semantics/OpSeq.re

```diff
@@ -17,6 +17,19 @@ let mk =
 let wrap = (operand: 'operand): t('operand, _) =>
   OpSeq(Placeholder(0), S(operand, E));
 
+let wrap_operator =
+    (
+      ~err: ErrStatus.t=NotInHole,
+      left: 'operand,
+      op: 'operator,
+      right: 'operand,
+    )
+    : t('operand, 'operator) =>
+  OpSeq(
+    Skel.(BinOp(err, op, Placeholder(0), Placeholder(0))),
```

Comment:
you shouldn't have two Placeholder(0)s in a BinOp...

Reply:
I don't really understand the meaning of these numbers, but I found a snippet in UHTyp that had the same code, but with a 1 in one of the placeholders, so I assume that's okay.

Reply:
Your fix is good 👍  

An `OpSeq.t` consists of a `Skel.t` and a `Seq.t`. The `Seq.t` is a flat sequential representation, which is good for editing. The `Skel.t` is the tree-structured representation determined by operator precedence, which is necessary for type-checking and evaluation. The `Placeholder` numbers in the `Skel.t` are indexes into the accompanying `Seq.t` so that you know which operand is at each skel leaf. For example, notice how `Statics.Exp.syn_skel` (which synthesizes a type for some part of opseq) recurses down the skel structure, then looks up the right operand in the accompanying seq when it hits `Placeholder`.

Reply:
Ahaa, thank you very much. So when combining 2 `OpSeq`s (or integrating one created like this in a full program), these placeholders should be renumbered?
Doesn't that take a lot of work when you insert something at the beginning of your program?

Reply:
For opseqs of length 1 or 2, like in this case, it's fine to manually number. But generally you should only construct the seq, then make a call to `OpSeq.mk` and pass in the right `Associator` function to handle the skel stuff for you. For example, take a look at `Action.Exp.mk_OpSeq` and how it is used. Any opseq construction you wanna do will take place in `Action`, so I would just make use of the existing `mk_OpSeq` functions there.

Reply:
Yeah, I wanted to use an `Associator` like that first, but that would result in a dependency cycle. The code would probably be simple if with just an `mk_OpSeq` argument or only return the `Seq`.